### PR TITLE
rustc_codegen_llvm: add support for writing summary bitcode

### PR DESCRIPTION
**Pull request by durin42**
Typical uses of ThinLTO don't have any use for this as a standalone file, but distributed ThinLTO uses this to make the linker phase more efficient. With clang you'd do something like `clang -flto=thin -fthin-link-bitcode=foo.indexing.o -c foo.c` and then get both foo.o (full of bitcode) and foo.indexing.o (just the summary or index part of the bitcode). That's then usable by a two-stage linking process that's more friendly to distributed build systems like bazel, which is why I'm working on this area.

I talked some to @teresajohnson about naming in this area, as things seem to be a little confused between various blog posts and build systems. "bitcode index" and "bitcode summary" tend to be a little too ambiguous, and she tends to use "thin link bitcode" and "minimized bitcode" (which matches the descriptions in LLVM). Since the clang option is thin-link-bitcode, I went with that to try and not add a new spelling in the world.

Per @dtolnay, you can work around the lack of this by using `lld --thinlto-index-only` to do the indexing on regular .o files of bitcode, but that is a bit wasteful on actions when we already have all the information in rustc and could just write out the matching minimized bitcode. I didn't test that at all in our infrastructure, because by the time I learned that I already had this patch largely written.

### CODE REVIEW

**Comment by Nadrieril** [comment]
r? compiler

**Comment by lcnr** [comment]
r? compiler

**Comment by durin42** [comment]
Latest CI run failed with `      [ui] tests/ui/asm/inline-syntax.rs#arm` - which I shouldn't have been able to influence at all with this change.

**Comment by nnethercote** [comment]
This is enabled with `--emit=thin-link-bitcode`, correct?

The code in the PR looks ok to me, though LTO experts might have opinions -- @bjorn3? @mw?

But I think that adding a new output type is a significant enough change that it should be done via the [Major Change Proposal](https://forge.rust-lang.org/compiler/mcp.html) mechanism.

**Comment by bjorn3** [comment]
> But I think that adding a new output type is a significant enough change that it should be done via the [Major Change Proposal](https://forge.rust-lang.org/compiler/mcp.html) mechanism.

https://github.com/rust-lang/compiler-team/issues/735

**Comment by nnethercote** [comment]
Ok, thanks for the extra info. @bjorn3 do you want to give the final approval here? You know what's going on here much better than I do.

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 1]
```diff
@@ -725,12 +728,31 @@ pub(crate) unsafe fn codegen(
...
+                }
+            }
+
             if config.emit_bc || config.emit_obj == EmitObj::Bitcode {
-                let _timer = cgcx
-                    .prof
-                    .generic_activity_with_arg("LLVM_module_codegen_emit_bitcode", &*module.name);
-                if let Err(err) = fs::write(&bc_out, data) {
-                    dcx.emit_err(WriteBytecode { path: &bc_out, err });
+                {
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/src/back/write.rs:748]: Why this extra block? It spans the entirety of the enclosing if statement.

**Comment by durin42** [comment]
I have confirmed that, somehow, the llvm-17 regression here is real and seemingly caused by my change. I'll investigate soon.

**Comment by durin42** [comment]
Fixed the llvm 17 failure. Sorry for the force-push, that ended up being hard to avoid once I got into the weeds debugging.

I also cleaned up a couple of variable names so things are generally consistent in using `summary` instead of `index`.

**Pull request merged**